File: tests/metrics/ulcer_index_test.py
from __future__ import annotations

import numpy as np
import pandas as pd
import pytest
import quantstats.stats as qs

from taflow import UlcerIndex as RollingUlcerIndex
from taflow.metrics import UlcerIndex as ExportedMetricUlcerIndex
from taflow.metrics.ulcer_index import UlcerIndex


def quantstats_ulcer_index(returns: np.ndarray) -> float:
    """Evaluate the pinned oracle with the DatetimeIndex it requires."""
    series = pd.Series(
        returns,
        index=pd.date_range("2000-01-01", periods=len(returns), freq="D"),
    )
    return float(qs.ulcer_index(series))


@pytest.mark.parametrize(
    "returns",
    [
        np.array([0.10, -0.20]),
        np.zeros(32),
        np.linspace(-0.04, 0.05, 101),
        np.array([0.15, -0.1, -0.1, 0.25, -0.3, 0.1]),
        np.array([0.2, -0.2, 0.25, -0.2, 0.25, -0.2]),
        np.random.default_rng(20260811).normal(0.0004, 0.012, 513),
    ],
)
def test_ulcer_index_matches_quantstats(returns: np.ndarray) -> None:
    actual = UlcerIndex.from_returns(returns).compute()
    expected = quantstats_ulcer_index(returns)
    assert actual == pytest.approx(expected, rel=1e-12, abs=1e-14)


def test_ulcer_index_factories_and_lifecycle_are_invariant() -> None:
    returns = np.array([0.10, -0.20, 0.05, -0.25, 0.1])
    expected = UlcerIndex.from_returns(returns).compute()
    equity = 100.0 * np.r_[1.0, np.cumprod(1.0 + returns)]
    pnl = np.diff(equity)

    assert UlcerIndex.from_equity(equity).compute() == pytest.approx(expected)
    assert UlcerIndex.from_pnl(pnl, initial_equity=100.0).compute() == pytest.approx(
        expected
    )
    assert UlcerIndex.from_log_returns(np.log1p(returns)).compute() == pytest.approx(
        expected
    )

    state = UlcerIndex.from_returns([])
    assert state.value is None
    assert state.append(returns[0]) is state
    assert state.value is None
    assert state.extend(returns[1:3]) is state
    assert state.extend(returns[3:]) is state
    assert state.compute() == pytest.approx(expected)
    assert len(state) == len(returns)
    assert state.reset() is state
    assert len(state) == 0
    assert state.value is None
    assert state.extend(returns).compute() == pytest.approx(expected)


def test_ulcer_index_uses_quantstats_phantom_wealth_and_divisor() -> None:
    # Drawdowns are 0, -0.2, 0, so sqrt(sum(dd**2) / (3 - 1)).
    expected = np.sqrt(0.04 / 2.0)
    assert UlcerIndex.from_returns([0.1, -0.2, 0.25]).compute() == pytest.approx(
        expected
    )
    assert UlcerIndex.from_returns([-0.1]).compute() is None


def test_ulcer_index_missing_and_invalid_values() -> None:
    filtered = np.array([0.1, -0.2, 0.05])
    state = UlcerIndex.from_returns([0.1, np.nan, -0.2, 0.05])
    assert len(state) == len(filtered)
    assert state.compute() == pytest.approx(quantstats_ulcer_index(filtered))
    with pytest.raises(ValueError):
        UlcerIndex.from_returns([np.nan], nan_policy="raise")
    with pytest.raises(ValueError):
        UlcerIndex.from_returns([np.inf])
    with pytest.raises(ValueError):
        UlcerIndex.from_returns([-1.01])


def test_ulcer_index_requires_semantic_factory() -> None:
    with pytest.raises(TypeError):
        UlcerIndex()


def test_whole_history_metric_namespace_is_distinct_from_rolling_indicator() -> None:
    assert ExportedMetricUlcerIndex is UlcerIndex
    assert RollingUlcerIndex is not UlcerIndex

File: crates/taflow-metrics/src/metrics/ulcer_index_test.rs
use super::ulcer_index::UlcerIndex;
use crate::{MetricInputKind, NanPolicy};

#[test]
fn matches_quantstats_phantom_wealth_and_sample_divisor() {
    let mut metric = UlcerIndex::new(NanPolicy::Omit).unwrap();
    metric.from_returns(&[]).unwrap();
    metric.extend(&[0.1, -0.2, 0.25]).unwrap();

    // Wealth is 1 -> 1.1 -> 0.88 -> 1.1 and drawdowns are 0, -0.2, 0.
    // QuantStats 0.0.81 divides the squared sum by returns.len() - 1.
    assert!((metric.value().unwrap() - (0.04_f64 / 2.0).sqrt()).abs() < 1e-15);
    assert_eq!(metric.compute(), metric.value());
    assert_eq!(metric.len(), 3);
}

#[test]
fn requires_two_usable_returns() {
    let mut metric = UlcerIndex::new(NanPolicy::Omit).unwrap();
    metric.from_returns(&[]).unwrap();
    assert_eq!(metric.value(), None);
    assert_eq!(metric.append(-0.1).unwrap(), None);
    // Both observations remain 10% below the phantom starting peak, and the
    // oracle divisor is n - 1, so sqrt((0.1^2 + 0.1^2) / 1).
    assert!((metric.append(0.0).unwrap().unwrap() - 0.02_f64.sqrt()).abs() < 1e-15);
}

#[test]
fn all_semantic_input_modes_have_the_same_result() {
    let returns = [0.1, -0.2, 0.05, -0.25, 0.1];
    let mut expected = UlcerIndex::new(NanPolicy::Omit).unwrap();
    expected.from_returns(&[]).unwrap();
    expected.extend(&returns).unwrap();

    let mut log_returns = UlcerIndex::new(NanPolicy::Omit).unwrap();
    log_returns.from_log_returns(&[]).unwrap();
    log_returns.extend(&returns.map(f64::ln_1p)).unwrap();

    let mut equity = UlcerIndex::new(NanPolicy::Omit).unwrap();
    equity.from_equity(&[]).unwrap();
    equity
        .extend(&[100.0, 110.0, 88.0, 92.4, 69.3, 76.23])
        .unwrap();

    let mut pnl = UlcerIndex::new(NanPolicy::Omit).unwrap();
    pnl.from_pnl(&[], 100.0).unwrap();
    pnl.extend(&[10.0, -22.0, 4.4, -23.1, 6.93]).unwrap();

    for actual in [log_returns.compute(), equity.compute(), pnl.compute()] {
        assert!((actual.unwrap() - expected.compute().unwrap()).abs() < 1e-14);
    }
}

#[test]
fn reset_replay_chunking_and_nan_omission_are_invariant() {
    let returns = [0.1, f64::NAN, -0.2, 0.05, -0.25, 0.1];
    let mut batch = UlcerIndex::new(NanPolicy::Omit).unwrap();
    batch.from_returns(&[]).unwrap();
    batch.extend(&returns).unwrap();

    let mut chunked = UlcerIndex::new(NanPolicy::Omit).unwrap();
    chunked.from_returns(&[]).unwrap();
    chunked.extend(&returns[..2]).unwrap();
    for &value in &returns[2..] {
        chunked.append(value).unwrap();
    }
    assert_eq!(chunked.compute(), batch.compute());
    assert_eq!(chunked.len(), 5);

    chunked.reset();
    assert!(chunked.is_empty());
    assert_eq!(chunked.value(), None);
    chunked.extend(&returns).unwrap();
    assert_eq!(chunked.compute(), batch.compute());
}

#[test]
fn rejects_non_path_semantic_domains() {
    let mut unbound = UlcerIndex::new(NanPolicy::Omit).unwrap();
    assert!(unbound.append(0.01).is_err());
}
